### PR TITLE
Scale, rotation and shift aware calibration and movement

### PR DESCRIPTION
**Pull request by astaff**
Per #397 

- Adds CLI tool for jogging the robot to calibration positions and
  calculating calibration parameters: `api/opentrons/cli/main.py`.
- Adds smoothie homing offsets for X and Y to avoid hitting endstops
  during calibration.
- `driver.move` now takes `target` a dictionary with axis values to
  move to.
- Adds coordinate rounding to 3rd digit before sending it to Smoothie
- Unifies motion dispatch and position update to fully utilise pose tracker
  refactoring `gantry.py` and consolidating Gantry, Mover, Actuator, Mount
  under Mover class. Plungers use mover too. Mover takes cartesian to driver
  coordinate mapping, coordinate system of origin and destination to.
  dispatch move commands to driver and update position.
- Refactors pipette and robot to use updated Mover.
- Utilizes `robot_config` to pull calibration values and offsets.
- Adds test for pickup accuracy.
- Adds calibration data for Avogadro and Franklin.

*TODO*

- README for CLI
- Tests for CLI

### CODE REVIEW

File: api/opentrons/commands/commands.py [hunk 1]
```diff
@@ -14,12 +14,12 @@ def make_command(name, payload):
     return {'name': name, 'payload': payload}
 
 
-def home(axis):
-    text = 'Homing pipette plunger on axis {axis}'.format(axis=axis)
+def home(mount):
```
Comment (mcous) [inline api/opentrons/commands/commands.py]: Don't disagree with this change, but it seems out of scope for this PR (as opposed to a PR that refactors homing and our use of axis / mount for pipettes)

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py [hunk 2]
```diff
@@ -136,11 +136,18 @@ def current(self):
...
-    def set_speed(self, value):
-        ''' set total movement speed in mm/second'''
-        speed = value * SEC_PER_MIN
-        command = GCODES['SET_SPEED'] + str(speed)
-        self._send_command(command)
+    def set_speed(self, values):
+        # reference: http://smoothieware.org/supported-g-codes
+        # “Push” the current feed-rate and seek-rate so that another one can
+        # be temporarily used, then the current one can be restored
+        self._send_command('M120')
```
Comment (mcous) [inline api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: Should this magic g-code string live in the `GCODES` dict?

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py [hunk 3]
```diff
@@ -44,7 +44,7 @@
           'PROBE': 'G38.2',
           'ABSOLUTE_COORDS': 'G90',
           'RESET_FROM_ERROR': 'M999',
-          'SET_SPEED': 'G0F',
+          'SET_SPEED': 'G0 {axis} F{value}',
```
Comment (mcous) [inline api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: 1. Is passing an axis here safe?

Seeing as `G0` is also move to, I think that passing an axis would be the same as saying "move {axis} to current position" (assuming G-code coordinates are modal), which scares me a little, even though it probably should have no effect.

2. Existing implementation would output `G0F{speed}`, whereas this adds spaces. Do spaces affect anything?

3. Rather than change how this g-codes dict works in this PR, I think it would be more sensible to have a separate PR that makes all these g-codes functional (given that there are only 10 of them)

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py [hunk 4]
```diff
@@ -136,11 +136,18 @@ def current(self):
     def speed(self):
         pass
 
-    def set_speed(self, value):
-        ''' set total movement speed in mm/second'''
-        speed = value * SEC_PER_MIN
-        command = GCODES['SET_SPEED'] + str(speed)
-        self._send_command(command)
+    def set_speed(self, values):
```
Comment (mcous) [inline api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: What is this set_speed now supposed to do? From reading this code and the smoothie documentation (and the [rep rap wiki's gcode page](http://reprap.org/wiki/G-code#M120:_Push)) it looks to me like calling `set_speed` is going to result in nothing changing:

1. `M120` - Save current configuration
2. `G0 {axis} F{rate}` - Change the configuration
3. Restore the saved configuration, wiping out any changes in (2)

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py [hunk 5]
```diff
@@ -172,6 +179,7 @@ def _send_command(self, command, timeout=None):
             if moving_plunger:
                 self.set_current('BC', PLUNGER_CURRENT_HIGH)
 
+            print(command)
```
Comment (mcous) [inline api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: Debugging statement should be removed

File: api/opentrons/api/calibration.py [hunk 6]
```diff
@@ -27,8 +27,8 @@ def _set_state(self, state):
 
     def tip_probe(self, instrument):
         self._set_state('probing')
-        calibration_functions.probe_instrument(
-            instrument._instrument, self._robot)
+        # calibration_functions.probe_instrument(
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/api/calibration.py]: debugging — fixed

File: api/opentrons/cli/main.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+# Expected reference points
+expected = [
+    (64.0, 92.8),       # dimple 4
+    (329.0, 92.8),      # dimple 6
+    (196.50, 273.80)    # dimple 11
+]
+
+# Actuals get overridden when ENTER is pressed
+actual = [
+    (33.0, 5.25),
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/cli/main.py]: replacing with `actual = [(0, 0)] * 3`

File: api/opentrons/commands/commands.py (hunk 1 above)
Comment (astaff) [reply api/opentrons/commands/commands.py]: bigger refactor of axis / mounts for sure. this one came around when i was working on 'Mover.home' as a part of the whole mover refactor.

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py (hunk 3 above)
Comment (astaff) [reply api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: I will retract this from PR. Speed setting is a different subject.

File: api/opentrons/cli/main.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
+#!/usr/bin/env python
```
Comment (mcous) [inline api/opentrons/cli/main.py]: [Food for thought, no change requested] Not sure where else to put this comment, but should these calibration tools live outside the API? Rather than being part of the API, I've imagined them more as (possibly privileged) users of the API

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py (hunk 4 above)
Comment (astaff) [reply api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: Please disregard. It should be a separate ticket and separate PR.

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py [hunk 9]
```diff
@@ -16,7 +16,7 @@
 
 # TODO (ben 2017112): figure out why fw setting wasn't working on Avogadro
 # and remove this
-HOMING_OFFSETS = 'M206 X5'
+HOMING_OFFSETS = 'M206 X5 Y5'
```
Comment (mcous) [inline api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: While we're touching this line, should we move `M206` to the gcodes dict?

File: api/opentrons/robot/gantry.py [hunk 10]
```diff
@@ -1,208 +1,112 @@
...
+                y=y + distance if axis == 'y' else y,
+                z=z + distance if axis == 'z' else z
+            )
+        )
 
-    # FIXME: Should instruments be able to set speed for overall system motion?
-    def set_speed(self, rate):
-        ''' Set combined motion speed '''
-        self.driver.set_speed(rate)
+    def move(self, pose, x, y=0, z=0):
```
Comment (mcous) [inline api/opentrons/robot/gantry.py]: Why are `y` and `z` defaulted to `0`? Also, why is `x` not defaulted to anything? Comment below seems to imply that they should all be defaulted to `None`

File: api/opentrons/robot/gantry.py [hunk 11]
```diff
@@ -1,208 +1,112 @@
...
+
+        # apply transformation
+        x, y, z = relative(
+            pose,
+            src=self._origin,
+            dst=self._reference,
+            src_point=Point(x, y, z))
+
+        target = Point(
+            x=x if 'x' in self._axis_mapping else 0,
```
Comment (mcous) [inline api/opentrons/robot/gantry.py]: Same question as above about `0` vs `None`. I feel like I'm misunderstanding something here, so I'm sorry if that's the case

File: api/opentrons/robot/robot_configs.py [hunk 12]
```diff
@@ -4,10 +4,10 @@
 
 from collections import namedtuple
 
-PLUNGER_CURRENT_LOW = '0.1'
-PLUNGER_CURRENT_HIGH = '0.5'
+PLUNGER_CURRENT_LOW = 0.1
+PLUNGER_CURRENT_HIGH = 0.5
 
-CURRENT_ROBOT = 'B2-4'
+CURRENT_ROBOT = 'B2-5'
```
Comment (mcous) [inline api/opentrons/robot/robot_configs.py]: [Not necessarily a change request] Depending on how long the true config ticket is going to take, we should move this to an env var like you suggested before we end up changing this line in every single PR

File: api/opentrons/robot/robot_configs.py [hunk 13]
```diff
@@ -28,52 +30,64 @@
...
+        [  1.88679245e-03,   9.97237569e-01,   0.00000000e+00,  -1.66290112e+00],    # NOQA
+        [  0.00000000e+00,   0.00000000e+00,   1.00000000e+00,   4.50000000e+00],    # NOQA
+        [ -5.03305613e-19,   2.60208521e-18,   0.00000000e+00,   1.00000000e+00]],   # NOQA
+    instrument_offset=(-40.13513347,  28.33733321,  -0.25)  # left to right
 )
 
 robots = {
     'B1': Ibn,
-    'B2-4': Amadeo,
+    'B2-4': Amedeo,
```
Comment (mcous) [inline api/opentrons/robot/robot_configs.py]: Well this is an amusing typo we introduced. Good fix!

File: api/setup.py [hunk 14]
```diff
@@ -32,7 +32,8 @@
 INSTALL_REQUIRES = [
     'pyserial==3.2.1',
     'aiohttp==2.2.3',
-    'numpy==1.13.0']
+    'numpy==1.13.0',
+    'urwid==1.3.1']
```
Comment (mcous) [inline api/setup.py]: If this is only for the CLI then please see my comment about calibration CLI living outside of API. If not, please diregard

File: api/opentrons/robot/robot.py [hunk 15]
```diff
@@ -164,6 +171,30 @@ def __init__(self):
         only once instance of a robot.
         """
         self._driver = driver_3_0.SmoothieDriver_3_0_0()
+
+        self._actuators = {
+            'left': {
+                'carriage': Mover(
```
Comment (btmorr) [inline api/opentrons/robot/robot.py]: I like the structure of this overall, but is carriage clear here? This is the first time I've seen it used.

File: api/opentrons/robot/robot.py [hunk 16]
```diff
@@ -233,13 +264,42 @@ def reset(self):
...
         )
-        self.poses = self.gantry._setup_mounts(self.poses)
+
+        # Extract only transformation component
+        inverse_transform = pose_tracker.inverse(
+            pose_tracker.transform(CALIBRATION))
+
+        self.poses = pose_tracker.bind(self.poses) \
+            .add(obj=id(CALIBRATION), transform=CALIBRATION) \
+            .add(obj=self.gantry, parent=id(CALIBRATION)) \
```
Comment (btmorr) [inline api/opentrons/robot/robot.py]: This way of accomplishing parentage makes the call stack for `move_head` particularly unclear

File: api/opentrons/robot/robot.py [hunk 17]
```diff
@@ -405,7 +475,6 @@ def home(self, *args, **kwargs):
     # TODO (ben 20171030): refactor use this to use public methods
     def move_head(self, *args, **kwargs):
         self.poses = self.gantry.move(self.poses, **kwargs)
-        self.poses = self.gantry._update_pose(self.poses)
```
Comment (btmorr) [inline api/opentrons/robot/robot.py]: Because of the parentage of gantry, I can no longer hand-trace the call stack for this method past `Mover.move`, and thus can't confirm that pose is being updated properly

File: api/opentrons/trackers/pose_tracker.py [hunk 18]
```diff
@@ -16,13 +16,27 @@ def __str__(self):
...
-        [0.0, 1.0, 0.0, y],
-        [0.0, 0.0, 1.0, z],
+        [1.0, 0.0, 0.0,   x],
+        [0.0, 1.0, 0.0,   y],
+        [0.0, 0.0, 1.0,   z],
         [0.0, 0.0, 0.0, 1.0]
     ])
 
 
+def transform(matrix) -> np.ndarray:
```
Comment (btmorr) [inline api/opentrons/trackers/pose_tracker.py]: This method ends up getting shadowed by pre-existing method parameters. Consider renaming

File: api/opentrons/trackers/pose_tracker.py [hunk 19]
```diff
@@ -129,15 +146,15 @@ def reverse(state, obj, root=None):
     ))
 
 
-def absolute(state, obj, root=None):
+def absolute(state, obj, root=None, point=Point(0, 0, 0)):
```
Comment (btmorr) [inline api/opentrons/trackers/pose_tracker.py]: The ability to specify this parameter makes it possible for the method to behave incorrectly. I'm opposed to this increase in power. Would prefer to add another transform representing the point, or make a vector comparison in the caller comparing against the returned value

File: api/opentrons/util/calibration_functions.py [hunk 20]
```diff
@@ -46,7 +40,8 @@ def probe_instrument(instrument, robot):
 
     *_, height = PROBE_SIZE
 
-    center = PROBE_TOP_COORDINATES - (0, 0, height)
+    center = PROBE_BOTTOM_COORDINATES + (0, 0, height / 2.0)
```
Comment (btmorr) [inline api/opentrons/util/calibration_functions.py]: What is the logic behind height/2?

File: api/opentrons/cli/main.py (hunk 8 above)
Comment (astaff) [reply api/opentrons/cli/main.py]: I see it being the like of `opentrons.server` and will keep it in `api` for now

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py (hunk 9 above)
Comment (astaff) [reply api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: this will likely go away once we have the whole homing situation figured out. HOMING_OFFSETS is descriptive enough IMO

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py [hunk 21]
```diff
@@ -192,9 +200,8 @@ def _setup(self):
     # ----------- END Private functions ----------- #
 
     # ----------- Public interface ---------------- #
-    def move(self, x=None, y=None, z=None, a=None, b=None, c=None):
+    def move(self, target):
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: Possibly

File: api/opentrons/robot/gantry.py (hunk 10 above)
Comment (astaff) [reply api/opentrons/robot/gantry.py]: For cases like plungers we need only one axis. I am also making callers pass all coordinates explicitly and handling None's / defaults upstream, like Pipette._move

File: api/opentrons/robot/gantry.py (hunk 11 above)
Comment (astaff) [reply api/opentrons/robot/gantry.py]: Great observation. That's a leftover from when the function was taking None's

File: api/opentrons/robot/robot.py [hunk 22]
```diff
@@ -265,8 +325,18 @@ def add_instrument(self, mount, instrument):
         to attach the instrument.
         """
         self._instruments[mount] = instrument
-        self.poses = self.gantry.mount_instrument(
-            self.poses, instrument, mount)
+        instrument.instrument_actuator = self._actuators[mount]['plunger']
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/robot/robot.py]: To keep this PR focused, we can clean up Pipette in the future refactors

File: api/opentrons/robot/robot_configs.py (hunk 12 above)
Comment (astaff) [reply api/opentrons/robot/robot_configs.py]: Agreed.

File: api/setup.py (hunk 14 above)
Comment (astaff) [reply api/setup.py]: That's for CLI. I don't anticipate CLI moving up in this very PR.

File: api/opentrons/api/calibration.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/api/calibration.py]: Will fix

File: api/opentrons/cli/main.py [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+# Actuals get overridden when ENTER is pressed
+actual = [
+    (33.0, 5.25),
+    (298.0, 6.25),
+    (169.5, 276.0),
+]
+# Accessible through 1,2,3 ... keyboard keys to test
+# calibration by moving to known locations
+test_points = [
+    (64.0, 92.8, TIP_LENGTH),           # dimple 4
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/cli/main.py]: Would this work?

```python
test_points = [(x, y, TIP_LENGTH) for x, y in expected] + \
    [
        (132.50, 90.50, TIP_LENGTH),        # Slot 5
        (332.13, 47.41, TIP_LENGTH),        # Corner of 3
        (190.65, 227.57, TIP_LENGTH),       # Corner of 8
        (330.14, 222.78, TIP_LENGTH),       # Corner of 9
    ]
```

File: api/opentrons/cli/main.py [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+    (190.65, 227.57, TIP_LENGTH),       # Corner of 8
+    (330.14, 222.78, TIP_LENGTH),       # Corner of 9
+]
+
+# World > Smoothie XY-plane transformation
+# Gets updated when you press SPACE after calibrating all points
+# You can also default it to the last known good calibration
+# if you want to test points or to measure real-world objects
+# using the tool
+XY = \
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/cli/main.py]: With no persistence, it's a convenience to have old calibration data there already, so you can go to test points and see how far off they are and what not.

File: api/opentrons/cli/main.py [hunk 25]
```diff
@@ -0,0 +1,249 @@
...
+    array([[  1.00094340e+00,  -3.45303867e-03,  -1.94897355e+01],    # NOQA
+           [  1.88679245e-03,   9.97237569e-01,  -1.66290112e+00],    # NOQA
+           [ -5.03305613e-19,   2.60208521e-18,   1.00000000e+00]])   # NOQA
+
+
+# Add fixed Z offset which is known so we don't have to calibrate for height
+# during calibration process
+def add_z(XY):
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/cli/main.py]: We are solving a 2D equation but the expected output is a 3D transformation. We need to store 2D matrix separately.

File: api/opentrons/cli/main.py [hunk 26]
```diff
@@ -0,0 +1,249 @@
...
+        'World: {0}'.format(tuple(dot(inv(T), list(current_position) + [1])[:-1])),  # NOQA
+        'Step: {0}'.format(step()),
+        'Current stage: ' + current_pipette,
+        'Message: ' + text
+    ])
+
+    status_text.set_text(text)
+
+
+def jog(axis, direction, step):
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/cli/main.py]: Gantry, robot, etc jog accepts  World coordinates. This takes smoothie.

File: api/opentrons/robot/robot.py (hunk 15 above)
Comment (astaff) [reply api/opentrons/robot/robot.py]: Got these names from Taylor

File: api/opentrons/cli/main.py [hunk 27]
```diff
@@ -0,0 +1,249 @@
...
+            status(repr(XY))
+    else:
+        try:
+            key_mappings[key]()
+        except KeyError:
+            status('invalid key: ' + repr(key))
+
+
+# move to test points based on current matrix value
+def validate(index):
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/cli/main.py]: In order to use these, calibration matrix needs to be loaded into robot. As it stands right now robot is hard to parametrize outside of module load.

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: I agree. So far that seems as a remedy to get it calibrate without hitting home switch

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: Agree. De-scoping set speed for now

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py [hunk 28]
```diff
@@ -244,11 +251,6 @@ def home(self, axis=AXES, disabled=DISABLE_AXES):
 
         position = HOMED_POSITION
 
-        if not self.simulating:
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: I agree. This particular one would make issue `M114` twice making it confusing to read G-code output while working on a ticket.

File: api/opentrons/instruments/pipette.py [hunk 29]
```diff
@@ -869,11 +889,14 @@ def home(self):
         <opentrons.instruments.pipette.Pipette object at ...>
         """
         @commands.publish.both(command=commands.home)
-        def _home(axis):
+        def home(mount):
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/instruments/pipette.py]: Good point. Fixed. Thanks!

File: api/opentrons/robot/gantry.py [hunk 30]
```diff
@@ -1,208 +1,112 @@
...
-        self.instrument = instrument
+from ..trackers.pose_tracker import (
+    Point, absolute, update, relative, ROOT
+)
+
+
+class Mover:
+    def __init__(
+        self,
+        *,
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/robot/gantry.py]: https://docs.python.org/3.5/reference/compound_stmts.html#function-definitions

```
Function call semantics are described in more detail in section Calls. A function call always assigns values to all parameters mentioned in the parameter list, either from position arguments, from keyword arguments, or from default values. If the form “*identifier” is present, it is initialized to a tuple receiving any excess positional parameters, defaulting to the empty tuple. If the form “**identifier” is present, it is initialized to a new dictionary receiving any excess keyword arguments, defaulting to a new empty dictionary. Parameters after “*” or “*identifier” are keyword-only parameters and may only be passed used keyword arguments.
```

An example from asyncio: https://docs.python.org/3/library/asyncio-stream.html

File: api/opentrons/robot/gantry.py [hunk 31]
```diff
@@ -1,208 +1,112 @@
...
 
-    # FIXME: Should instruments be able to set speed for overall system motion?
-    def set_speed(self, rate):
-        ''' Set combined motion speed '''
-        self.driver.set_speed(rate)
+    def move(self, pose, x, y=0, z=0):
+        # Default coordinates that are None to current so we can
+        # pass them around safely
+        if self._parent:
+            pose = self._parent.move(pose, x, y, z)
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/robot/gantry.py]: Movers are nested. For example, pipette is a child of gantry and moving pipette means moving gantry and then pipette

File: api/opentrons/robot/gantry.py [hunk 32]
```diff
@@ -1,208 +1,112 @@
...
+        target = Point(
+            x=x if 'x' in self._axis_mapping else 0,
+            y=y if 'y' in self._axis_mapping else 0,
+            z=z if 'z' in self._axis_mapping else 0)
+
+        # driver axis to point axis value mapping
+        # NOTE: point._asdict() returns point as dictionary:
+        # {'x': x, 'y': y, 'z': z}
+        driver_target = {
+            driver_axis: target._asdict()[xyz]
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/robot/gantry.py]: How do you feel about this one:

```python
        if 'x' in self._axis_mapping:
            driver_target[self._axis_mapping['x']] = x

        if 'y' in self._axis_mapping:
            driver_target[self._axis_mapping['y']] = y

        if 'z' in self._axis_mapping:
            driver_target[self._axis_mapping['z']] = z
```

File: api/opentrons/robot/gantry.py [hunk 33]
```diff
@@ -1,208 +1,112 @@
...
+        return update(pose, self, point)
 
-    def __init__(self, gantry, mount, instrument):
-        self.gantry = gantry
-        self.mount = mount
-        self.instrument = instrument
+    def set_speed(self, value):
+        pass
+        # TODO (artyom 20171105): uncomment once proper plunger speeds are
+        # defined
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/robot/gantry.py]: Need to create a separate ticket for handling speeds correctly. Right now this is not being done.

File: api/opentrons/trackers/pose_tracker.py (hunk 18 above)
Comment (astaff) [reply api/opentrons/trackers/pose_tracker.py]: `extract_transform`

File: api/opentrons/trackers/pose_tracker.py (hunk 19 above)
Comment (astaff) [reply api/opentrons/trackers/pose_tracker.py]: Can you give an example of incorrect behavior?

File: api/opentrons/util/calibration_functions.py (hunk 20 above)
Comment (astaff) [reply api/opentrons/util/calibration_functions.py]: we are defining center as XY of the probe and half of it's height

File: api/opentrons/robot/robot.py (hunk 16 above)
Comment (astaff) [reply api/opentrons/robot/robot.py]: Not sure I understand this comment

File: api/opentrons/robot/gantry.py [hunk 34]
```diff
@@ -1,208 +1,112 @@
...
+        origin=ROOT,
+        children=None
+    ):
+        self._driver = driver
+        self._origin = origin
+        self._reference = driver if reference is None else reference
+        self._parent = None
+        self._axis_mapping = axis_mapping
+        self._children = children or []
+        [child.attach(self) for child in self._children]
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/robot/gantry.py]: I agree. Ideally this should re-use tree-like structure created for pose tracker and become functional. Where it stands right now, we need children to home (homing parent will home all of it's children). I would suggest to open a ticket to make it functional and merge as is for now.

File: api/opentrons/robot/gantry.py [hunk 35]
```diff
@@ -1,208 +1,116 @@
...
-    def __init__(self, driver, axis, instrument):
-        self.axis = axis
-        self.driver = driver
-        self.instrument = instrument
+from ..trackers.pose_tracker import (
+    Point, absolute, update, relative, ROOT
+)
+
+
+class Mover:
```
Comment (mcous) [inline api/opentrons/robot/gantry.py]: I thought I put this comment in a previous review, but I guess I forgot to hit enter or something. Can you give me a quick summary of how this class was designed and what it accomplishes? Between this new class and the `_actuators` changes in `robot` I'm finding the system a bit hard to track in my head

Comment (astaff) [reply api/opentrons/robot/gantry.py]: This class translates movements from `origin`relative to `reference`, takes mappings from cartesian axes and driver's axes and dispatches movement calls to driver, updating pose tree accordingly.

Movers are organized hierarchically, and move calls to children are being dispatched to parents first. For example Instrument's mover will call Gantry's mover.

I agree Movers, Actuators situation is pretty confusing and we need to get to the common terminology. Robot's _actuators is the attempt to start that.

With single mover, that can be bound to origin, coordinate mappings and reference, we can avoid repeating ourselves as it was the case with Gantry and InstrumentMover before. This also helps to keep construction of Movers tree centralized.

What are the particular parts that are hard to track? I'll make sure to address that.

File: api/opentrons/robot/gantry.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/robot/gantry.py]: Also this PEP: https://www.python.org/dev/peps/pep-3102/
I would appreciate your suggestions to enforce named arguments in any other ways, @mcous

[the remark this answers is not in the document]
Comment (mcous) [reply api/opentrons/robot/gantry.py]: > I would appreciate your suggestions to enforce named arguments

Why is it important to enforce named arguments here, especially if it comes at the expense of readability? Just because there's a PEP for it doesn't mean we should use it, especially if it appears nowhere else (that I know of) in our codebase. Consistency and readability win every single time for me.

Also, for future reference with these sorts of comments: it took a lot of reading of Python docs to get to the conclusion "in Python 3, any arguments after `*` are only able to be used as named arguments" (your text box was cut off and your own comments didn't mention named arguments at all until your last comment).

[the remark this answers is not in the document]
Comment (btmorr) [reply api/opentrons/robot/gantry.py]: Spoke w/ @astaff f2f, going to remove this and remember to use named arguments at construction site

File: api/opentrons/robot/gantry.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (btmorr) [reply api/opentrons/robot/gantry.py]: Spoke to @astaff f2f, tree traversal for children is not needed at this point. It was primarily for homing sequence, but homing sequence should be ordered by hand as it is currently designed to avoid collision or accidental aspiration. Doubly-linked tree not currently needed.

File: api/opentrons/robot/gantry.py (hunk 10 above)
Comment (btmorr) [reply api/opentrons/robot/gantry.py]: f2f: changing all defaults to None

File: api/opentrons/cli/main.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/cli/main.py]: As per f2f with @btmorr now it is
```python
array([
        [ 1.00, 0.00, -35.23],   # NOQA
        [-0.01, 1.00,  -1.81],   # NOQA
        [ 0.00, 0.00,   1.00]])  # NOQA
```

File: api/opentrons/robot/gantry.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/robot/gantry.py]: Removed

File: api/opentrons/robot/robot.py (hunk 17 above)
Comment (astaff) [reply api/opentrons/robot/robot.py]: Removed parent from mover

File: api/opentrons/robot/gantry.py (hunk 10 above)
Comment (astaff) [reply api/opentrons/robot/gantry.py]: Done

File: api/opentrons/robot/robot_configs.py [hunk 36]
```diff
@@ -28,52 +30,71 @@
...
-        PLUNGER_CURRENT_LOW),
-    deck_offset=(-31.45, -20.1, 0),
-    probe_center={'z': 57.81, 'x': 259.8, 'y': 298.875},
-    probe_dimensions={'length': 41, 'width': 38.7, 'height': 67.81}
+    current='M907 X1.2 Y1.5 Z0.8 A0.8 B{0} C{0}'.format(PLUNGER_CURRENT_LOW),
+    deck_offset=None,
+    probe_center=(287, 295, 55.0),
+    # X, Y and Z measurement of imaginary bounding box surrounding the probe
+    # giving safe distance to position for probing
+    probe_dimensions=(50.0, 50.0, 53.0),
```
Comment (btmorr) [inline api/opentrons/robot/robot_configs.py]: Just noticed this. Why did these values change for Rosalind?

File: api/opentrons/trackers/pose_tracker.py (hunk 19 above)
Comment (btmorr) [reply api/opentrons/trackers/pose_tracker.py]: This makes the function more general, allowing it to be used to find the "absolute" position of the object with respect to a non-origin point. This is the behavior that we are referring to as "relative" position. It muddies the terms. Also, the additional term in the `relative` function makes it un-intuitive how to use both src and src_point.

File: api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (btmorr) [reply api/opentrons/drivers/smoothie_drivers/v3_0_0/driver_3_0.py]: Bump

File: api/opentrons/robot/mover.py [hunk 37]
```diff
@@ -0,0 +1,96 @@
+from ..trackers.pose_tracker import Point, change_base, update, ROOT
+
+
+class Mover:
+    def __init__(self, driver, axis_mapping, dst, src=ROOT):
+        self._driver = driver
+        self._axis_mapping = axis_mapping
+        self._dst = dst
```
Comment (mcous) [inline api/opentrons/robot/mover.py]: [Nitpick, no change requested]: generally, I prefer fully spelled out words (or, partially spelled words without dropped vowels e.g. `dest`). These are pretty obvious, so whatever, but yeah

**Comment by btmorr** [comment]
Documenting f2f 👍 from @mcous
